### PR TITLE
v1.15.2 Include VersionInfo in Scaladoc

### PR DESCRIPTION
**Pull request by jdcba**
1. Use `val` in VersionInfoPlugin.scala — avoid repeated invocations of `git` CLI per build.
2. Include VersionInfo in generated Scaladoc (`sbt doc`) — improve developer API reference.

Before:

screen shot 2018-01-16 at 3 30 47 pm

After:

screen shot 2018-01-16 at 3 33 18 pm

### CODE REVIEW

File: uniform-core/src/main/scala/au/com/cba/omnia/uniform/core/version/VersionInfoPlugin.scala [hunk 1]
```diff
@@ -25,29 +25,36 @@ object VersionInfoPlugin extends Plugin {
...
 
       IO.write(
-        txt,
-        s"""VERSION=$version
-           |GIT=${commit(base).show}
-           |DATE=${timestamp(now)}""".stripMargin
+        txtFile,
+        s"""VERSION=${version}
+           |GIT=${git}
+           |DATE=${date}""".stripMargin
```
Comment (toddmowen) [inline uniform-core/src/main/scala/au/com/cba/omnia/uniform/core/version/VersionInfoPlugin.scala]: Should `${verbose}` be written to the the text file too, for consistency?

**Comment by toddmowen** [comment]
Thinking about this a bit more, the timestamp needs to be computed at a high enough level for SBT itself to use it, otherwise it won't be reflected in the names of the published artifacts. Buried in this function is not the right place for it.

Seems like this whole VersionInfo thing might have been an early attempt to implement "reproducible builds", that actually turned out to be a dead end.

File: uniform-core/src/main/scala/au/com/cba/omnia/uniform/core/version/VersionInfoPlugin.scala (hunk 1 above)
Comment (jdcba) [reply uniform-core/src/main/scala/au/com/cba/omnia/uniform/core/version/VersionInfoPlugin.scala]: I don’t have an opinion but happy to add it in if you prefer.

**Comment by jdcba** [comment]
Hi @toddmowen this file serves a slightly different purpose than the CI package versions. Currently, VersionInfo is the only way for Scala code to detect/display/log/email its own version or git commit hash. This patch is only a bugfix from the missing Scaladoc, it is unopinionated about the values themselves.

VersionInfo is generated for all builds regardless of whether they are local or CI based. In trying to cover all scenarios it is like a lowest common denominator: not well specialised for any individual case, but not wrong either. In your example of Houston 2.0.4:

```
scala> import au.com.cba.omnia.houston.common.VersionInfo

scala> VersionInfo.version
res0: String = 2.0.4-20170717000007-7a5887b

scala> VersionInfo.date
res1: String = 20170717000237

scala> VersionInfo.git
res2: String = 7a5887b98d00d25806e710fa9917a45d9de3fb11-SNAPSHOT

scala> VersionInfo.verbose
res3: String = 2.0.4-20170717000007-7a5887b-20170717000237-7a5887b

```

As you can see, the verbose info is not wrong, it is simply (as it says in the name) verbose.

Re your comment about calculating the timestamp at a higher level. Currently, the master package timestamp is overwritten into `version.sbt` at the top level of our CI tooling, but only for CI builds, not for `localVersionSettings` builds. This is because the package string needs to be stable across all invocations of all tools in the build. In contrast, the verbose timestamp added by VersionInfo is specific to sbt and SNAPSHOTs.

**Comment by toddmowen** [comment]
```
scala> VersionInfo.verbose
res3: String = 2.0.4-20170717000007-7a5887b-20170717000237-7a5887b
```

The repetition of the commit hash is redundant, the inclusion of two separate timestamps is confusing, and what use does the second timestamp serve?

To avoid debate over the definition of the word "wrong", let me instead assert that the additional timestamp and commit hash have no practical use, and therefore shouldn't be included.

> This patch is only a bugfix for the missing Scaladoc.

But it doesn't actually document what the strings represent, hence the current discussion.

**Comment by jdcba** [comment]
Yes I can add some documentation to the Scaladoc. The extra timestamp is not useful for releases, only snapshots. Because the duplication is not always there, some dynamic logic would be required to reduce it.

**Comment by jdcba** [comment]
Added documentation. Mostly just repeating what’s already in `uniform` code, but with speculative references to the common practices of user projects (not part of `uniform` per se):

screen shot 2018-01-17 at 3 36 23 pm

**Comment by jdcba** [comment]
Yeah basically the three places cater for this issue in different scenarios:

* Uniform: lowest common denominator, works with or without the presence of CI tooling. Needed when developing locally and when online tooling is unavailable.
* TeamCity: when CI tooling concerns are ‘meta’ (external to) code repos.
* CI/CI2: when CI tooling concerns are ‘coupled’ (internal to) code repos.

The real world is a mixture of these things (each of the above models is individually flawed and incomplete).

**Pull request merged**